adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -0,0 +1,379 @@
+        });
+    }
+
+    private void doUploadMessage(final AmqpContext context, final Future<MessageSender> senderFuture,
+            final String endpointName) {
+
+        final Future<JsonObject> tokenFuture = getRegistrationAssertion(context.getTenantId(), context.getDeviceId(),
+                null);
+        final Future<TenantObject> tenantConfigFuture = getTenantConfiguration(context.getTenantId());
+
+        CompositeFuture.all(tenantConfigFuture, tokenFuture, senderFuture).compose(ok -> {
+            TenantObject tenantObject = tenantConfigFuture.result();
+            if (tenantObject.isAdapterEnabled(context.getTenantId())) {
+                final MessageSender sender = senderFuture.result();
+                final Message downstreamMessage = newMessage(context.getResourceIdentifier(),
+                        sender.isRegistrationAssertionRequired(),
+                        endpointName, context.getMessageContentType(), context.getMessagePayload(),
+                        tokenFuture.result(), null);
+
+                if (context.delivery().remotelySettled()) {
+                    // client uses AT_MOST_ONCE delivery semantics -> fire and forget
+                    return sender.send(downstreamMessage);
+                } else {
+                    // client uses AT_LEAST_ONCE delivery semantics
+                    return sender.sendAndWaitForOutcome(downstreamMessage);
+                }
+            } else {
+                // this adapter is not enabled for tenant
+                return Future.failedFuture(
+                        new AmqpErrorException(AmqpError.NOT_ALLOWED, "This adapter is not enabled for tenant."));
+            }
+        }).compose(downstreamDelivery -> {
+            LOG.info("successfully processed message [target: {}, QoS: {}] for device [tenantId: {}, deviceId: {}]");
+            if (context.delivery().remotelySettled()) {
+                // client uses AT_MOST_ONCE delivery semantics
+                // accept & settle the message regardless of whether
+                // the downstream peer accepted/rejected the message.
+                ProtonHelper.accepted(context.delivery(), true);
+            } else {
+                // client uses AT_LEAST_ONCE delivery semantics
+                // forward disposition received from downstream peer back to the client device.
+                LOG.info("AT_LEAST_ONCE delivery semantics");
+                context.delivery().disposition(downstreamDelivery.getRemoteState(),
+                        downstreamDelivery.remotelySettled());
+            }
+            return Future.<Void> succeededFuture();
+        }).recover(t -> {
+            LOG.info("Cannot process message for [deviceId: {}] of [tenantId: {}]", context.getDeviceId(),
+                    context.getTenantId());
+            if (AmqpErrorException.class.isInstance(t)) {
+                AmqpErrorException error = (AmqpErrorException) t;
+                context.reject(error);
```

Comment:
I like the idea of encapsulating the message disposition handling within the AmqpContext class. However, IMHO you should then either move all of the disposition handling code into this class or have that class provide methods for accepting, releasing and sending disposition frames as well

Reply:
yes